Add new `cardano-wallet-integration:e2e` tests running on preprod

Anviking opened the pull request:
- [x] Add a new `cardano-wallet-integration:e2e` test suite that:
    - has, and intends to have, a minimal number of tests. It does not want to duplicate everything from `cardano-wallet-integration`.
    - mostly uses the same `Test.Integration.Framework.DSL` the integration tests are written in
    - is replacing `lib/wallet-e2e` and is intended to soon replace the ruby e2e tests
- [x] Run this `cardano-wallet-integration:e2e` in CI on windows
    - [new run which hopefully passes like the prev one](https://buildkite.com/cardano-foundation/cardano-wallet/builds/9261#0194d089-3ef3-4f12-85ac-ad68f5a5f0b7) and should now make the stages of the setup appear as sections in the buildkite logs
- [x] No longer run ruby e2e tests on win

### Global preprod setup

We use only two wallets that are funded with ada. The transaction test will send funds from the richer to the poorer wallet. This should reduce the need for manual involvement as the funds last longer, and in a pinch we could even get by funding just one of the test wallets.

### How to run

```bash
export HAL_E2E_PREPROD_MNEMONICS='<mnemonics be found in our vault>'
cabal test cardano-wallet-integration:e2e
```

With node and wallet stdout logs manually cleaned:

```bash
transactions
  simple transaction (TRANS_CREATE_01x) [✔]
stake pools
  list
    some have non-zero rewards, stake, producedBlocks, saturation [✔]
    some have metadata [‐]
      # PENDING: metadata fetching not configured; could be enabled
```

### Issue Number

#4931

Anviking committed:
Add new `cardano-wallet-integration:e2e` tests

These run on preprod and are to replace both the `wallet-e2e` and ruby
e2e tests.

Anviking committed:
Delete `lib/wallet-e2e`

Anviking committed:
Decrease cardano-node log output verbosity

This makes it feasible to have both the node and wallet log to stdout in
the cardano-wallet-integration:e2e tests.

Anviking commented on .buildkite/pipeline.yml:
```diff
@@ -587,15 +587,20 @@ steps:
+        E2E_CONFIG_DIR: "configs\\cardano\\preprod"
         TESTS_E2E_FIXTURES: "$FIXTURE_DECRYPTION_KEY"
-        NODE_DB_DIR: "test\\e2e\\state\\node_db\\preprod"
+        NODE_DB_DIR: "lib\\wallet-e2e\\test-state\\preprod\\node\\db"
```
Set twice 😰 

```suggestion
```

Anviking committed:
fixup: drop now redundant env var in pipeline

Anviking commented on justfile:
```diff
@@ -72,31 +72,13 @@ unit-tests-local-cluster-match match:
-    -c lib/wallet-e2e/config/cardano-node/preprod \
-    -t lib/wallet-e2e/test-output/preprod
+# run cardano-wallet-integration:e2e suite against the preprod network
+e2e:
```
This doesn't work, will have a look

Anviking replied:
Now working.

Anviking commented on lib/integration/exe/e2e.hs:
```diff
@@ -0,0 +1,221 @@
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE RecordWildCards #-}
+{-# LANGUAGE TypeApplications #-}
+
```
```suggestion
```

Anviking commented on lib/integration/framework/Test/Integration/Framework/DSL.hs:
```diff
@@ -2295,8 +2406,15 @@ fixtureWalletWithMnemonics nextWalletMnemonic ctx label = do
+                (Link.listAddresses @'Shelley w) Default Empty
+    liftIO $ print addrs
+
+    liftIO $ race (threadDelay (6000 * oneSecond)) (checkBalance w) >>= \case
```
Oops, should revert this

Anviking commented on lib/local-cluster/lib/Cardano/Node/Cli/Launcher.hs:
```diff
@@ -64,9 +68,9 @@ stop (NodeInstance process) = do
+newtype NodeApi = NodeApi String
 
-nodeApiSocket :: NodeApi -> AbsFile
+nodeApiSocket :: NodeApi -> String
```
Hm, this should most likely be reverted

Anviking commented on lib/integration/exe/e2e.hs:
```diff
@@ -0,0 +1,221 @@
+    walletDbDir <- fromMaybe defaultWalletDbDir <$> lookupEnv "WALLET_DB_DIR"
+    nodeDbDir <- fromMaybe defaultNodeDbDir <$> lookupEnv "NODE_DB_DIR"
+    nodeDir <- fromMaybe defaultNodeDir <$> lookupEnv "NODE_DIR"
+    configDir <- fromMaybe defaultConfigDir <$> lookupEnv "E2E_CONFIG_DIR"
```
Would be nicer if we 1) embedded the mithril setup, which I think was the idea 2) maybe enbeded the config, and then maybe ended up with a CLI like this:

```bash
export HAL_E2E_PREPROD_MNEMONICS=<...> 
cardano-wallet-e2e --mitrhil # no other config besides the mnemonics and executables in path
cardano-wallet-e2e --state-dir # for dev, could be used for machine-global state in ci
cardano-wallet-e2e --wallet-url localhost:8090 # for dev / even faster iteration if you launch node & wallet yourself first, but arguably the `--state-dir` would already be fast enough

other global options:
--config # override the embedded config
```

paolino replied:
As long there is a working script which is machine tested by the CI (with some buildkite noise guarded by the `BUILDKITE` env var presence) I think we are fine. Also `just` has a decent calling interface and should work on linux and macos.
Making sure processes are always removed is not nice to see in the script, but scripts are much better in terms of documenting effort.
Controlling the processes from the executable is something nice which we should support for the after-daedalus era but I think it requires some design effort.

Anviking replied:
> As long there is a working script which is machine tested by the CI (with some buildkite noise guarded by the BUILDKITE env var presence) I think we are fine.

👍 

> Making sure processes are always removed is not nice to see in the script, but scripts are much better in terms of documenting effort.
> Controlling the processes from the executable is something nice which we should support for the after-daedalus era but I think it requires some design effort.

I don't completely follow this part — these e2e tests are currently launching `cardano-node` and `cardano-wallet`. I imagine it's a small step to let it also handle `mithril`, and then we don't need to coordinate paths like NODE_DB_DIR or NODE_DIR between the haskell code and the wrapping script, so hope this will work 🤞

paolino commented on cabal.project:
```diff
@@ -89,7 +89,6 @@ packages:
   lib/wai-middleware-logging/
   lib/wallet-benchmarks/
   lib/wallet/
-  lib/wallet-e2e/
```
:sad:

paolino commented on scripts/buildkite/main/windows-e2e.bat:
```diff
@@ -24,21 +24,11 @@ ls %NODE_DB_DIR%
-call bundle exec rake stop_node_and_wallet[preprod]
+echo %NODE_DB_DIR%
+echo %WALLET_DB_DIR%
+test\e2e\bins\cardano-wallet-integration-test-e2e.exe
```
:+1:

paolino commented on configs/cardano/preprod/config.json:
This is a problem. Next time we run the download script we lose your changes. Maybe move this patch into the integration setup ?

Anviking replied:
https://github.com/cardano-foundation/cardano-wallet/pull/4949/commits/b0de372ab52b946162e5a458bfe28142d9e122e2

Anviking replied:
Or rather: https://github.com/cardano-foundation/cardano-wallet/pull/4949/commits/90c4a525e52f029dd6bfd9548c592fc570b5ae76

Unfortunately I had to duplicate the preprod configs, but they can still be refreshed without reverting the reduced log verbosity, and maybe this could be fixed later.

paolino commented on .buildkite/pipeline.yml:
Maybe add the linux and macos (for both arch) steps as well ?

paolino replied:
Not as an alternative to the ruby ones, for now.

Anviking replied:
I'd favor doing it in a follow up PR.

Btw I presume you meant "<s>Not</s> **As** an alternative".

paolino replied:
Well I actually meant "not as an alternative" because removing the ruby one will be a slow and painful process during which both versions of E2E will be there

Anviking committed:
Embed & tweak configs for e2e tests

- Add `withConfigs` function that reads configs at compile time, writes
  them to tmp dir with tweaks to tracer verbosity at runtime.

- We no longer have to set the E2E_CONFIG_DIR env var

- Restore the previous tracer verbosities in configs/cardano/preprod/config.json

- Unfortunately we need to duplicate the configs to make it build with
  haskell.nix. Neither "../.." paths, or "ln -s ../../configs configs"
  appeared to work.

paolino reviewed (approved):
thanks!

Anviking committed:
Test that the duplicated configs don't diverge

paolino commented on .buildkite/pipeline.yml:
```diff
@@ -583,6 +583,7 @@ steps:
       depends_on:
         - trigger-windows-e2e-tests
       commands:
+        - diff -r configs/cardano/preprod lib/integration/configs/cardano/preprod # ensure the configs don't diverge despite being duplicated
```
:+1:
